LanguageFeatures/Patterns/matching_record_A01_t01.dart

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion At runtime, a pattern is matched against a value. This determines
+/// whether or not the match fails and the pattern refutes the value. If the
+/// match succeeds, the pattern may also destructure data from the object or
+/// bind variables.
+///
+/// Refutable patterns usually occur in a context where match refutation causes
+/// execution to skip over the body of code where any variables bound by the
+/// pattern are in scope. If a pattern match failure occurs in an irrefutable
+/// context, a runtime error is thrown.
+///
+/// To match a pattern p against a value v:
+/// ...
+/// Record:
+/// i. If the runtime type of v is not a subtype of the required type of p, then
+///   the match fails.
+/// ii. For each field f in p, in source order:
+///   a. Access the corresponding field in record v as r.
+///   b. Match the subpattern of f against r. If the match fails, the record
+///     match fails.
+/// iii. The match succeeds if all field subpatterns match.
+///
+/// @description Checks that if the runtime type of `v` is not a subtype of the
+/// required type of `p` then the match fails
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/expect.dart";
+
+String test1(Object o) {
+  switch (o) {
+    case (1, 2):
+      return "match-1";
+    case (3.14, name: String v1):
+      return "match-2";
+    case (value: 42, name: String v1):
+      return "match-3";
+    default:
+      return "no match";
+  }
+}
+
+String test2(Object o) {
+  if (o case (1, 2)) {
+    return "match-1";
+  }
+  if (o case (3.14, name: String v1)) {
+    return "match-2";
+  }
+  if (o case (value: 42, name: String v1)) {
+    return "match-3";
+  }
+  return "no match";
+}
+
+String test3(Object o) {
+  return switch (o) {
+      (1, 2) => "match-1",
+      (3.14, name: String v1) => "match-2",
+      (value: 42, name: String v1) => "match-3",
+      _ => "no match"
+    };
+}
+
+void test4(dynamic o) {
+  var (int v1, num _) = o;
+}
+
+main() {
+  Expect.equals("no match", test1(()));
+  Expect.equals("no match", test1((1,)));
+  Expect.equals("no match", test1((1, "2")));
```

Comment:
Considering the description, 

> if the runtime type of `v` is not a subtype of the
> required type of `p` then the match fails

this is only about the shape test (because the required type of a record pattern is always `Object?` at every field, in some overall shape). In order to actually test the shape check we should probably have a subpattern for each field which is guaranteed to match. An easy choice would be `_` in every field.

We could of course also interpret the description as applicable recursively (so we'd also have subpatterns, and those subpatterns and their matching behavior should also be tested). In that case we could use `T _` where `T` is some appropriate type.

There would presumably be some other test where the description says that we're testing the matching step for each of the fields of the record according to a constant (that is, we're testing their value, not the required type).

(Surely we already have that kind of test, named something like `...constant...', right? In that case we should be able to reduce this one such that it tests required types and match failure.)

Reply:
Thanks again, looks good now!